### ISSUE TITLE
CI: simplify tests

### ISSUE THREAD

**Issue by clason** [nvim-treesitter/nvim-treesitter#2313]
With the increasing number of parsers (and more coming), CI is starting to take a while...

We should pare down the number of tests:

1. reduce the combinations of platform, compiler, and Neovim version;
2. only run one combination (Ubuntu, gcc, nightly? that seems to be the fastest) on draft PRs (and push people to keep the PRs as draft until actually ready for review);

~3. avoid running all checks on new PRs twice (once for open, once for push)~ see #2314

We could also look into using labels for controlling workflows (like `ci-benchmark`, or `ci-full`, or `ci-none` -- the automated "Update README" PR for example does not need any tests).

**Comment by vigoux** [comment]
This all seems like a great idea, do you have exemples about controlling CI based on labels somewhere ?

**Comment by clason** [comment]
Not off the top of my head, but Homebrew uses that.

**Comment by MunifTanjim** [comment]
Compiled parser can be cached based on `lockfile.json` hash and only recompile when `lockfile.json` changes. That would cut down significant amount of time (4-5 minutes on each `workflows/tests.yml` run).

**Comment by clason** [comment]
Nice, we should do that then. Do you know how? 😄

**Comment by MunifTanjim** [comment]
I'll open a PR soon. 😄

**Comment by clason** [comment]
I'll merge my PR shortly, once tree-sitter releases the new version. (I can also merge it now if that helps.)

**Comment by theHamsta** [comment]
@MunifTanjim the relevant files are lockfile.json and `parsers.lua`. The compilation should only happen when those files change or we only compile on one config when those files have not changed. Could also do ccache caching on Ubunut/MacOS (I planned also caching for compiling the Rust dependency for the tests)

**Comment by theHamsta** [comment]
@clason it would be interesting not to fail when people have the old tree-sitter CLI when we do the ABI upgrade.

**Comment by clason** [comment]
Yes, that should be possible if we do the version check (only specify ABI if tree-sitter CLI is new enough).

**Comment by clason** [comment]
then approve and merge plz :]

**Comment by theHamsta** [comment]
@MunifTanjim there are workflows like this which let you test whether specific files changed similar to Gitlabs if changed option https://github.com/tj-actions/changed-files
 
and ... I forgot about `install.lua`. That should also trigger a recompilatoin

**Comment by clason** [comment]
Another thing we could think about is running the lockfile update PR more or less regularly (say, once daily) instead of every push to master -- which could be either too often or too seldom, depending on our activity...

**Comment by theHamsta** [comment]
We had this idea earlier and we wanted to actually align the lockfile updates with our activity. Until now it only does a git request (not even checkout)  which is not expensive. This would change with @vigoux PR. I would have just grepped the parser.c for the ABI version (and by that also realizing when there's actually no ABI present)

**Comment by clason** [comment]
I understand, but -- correct me if I'm wrong -- it mostly relies on _external_ changes? Which does not align with our activity.

(But I guess you were concerned with the situation where you adapted queries to changed parsers, and you wanted to simplify updating the parsers when you were done?)

**Comment by theHamsta** [comment]
In the ABI request, you'd test what users have and see whether they'd experience breakages. After that PR, we can upgrade. Neovim could be done even now.

**Comment by dundargoc** [comment]
> the automated "Update README" PR for example does not need any tests).

Tests are only run on the automated PR because they're being closed and reopened manually by someone. Is there a reason for this?

**Comment by clason** [comment]
Workflows started by CI cannot start new workflows themselves (AFAIU). So when the "Update Parser" workflow is started, the tests are not; I need to close and reopen the PR to trigger them.

**Comment by dundargoc** [comment]
Aha, so it's needed for the "Updated Parser" PR but not "Update README" PR? Gotcha.

For the "Update README" PRs the simplest way to skip tests would be to not closing and reopening them (unless there are other workflows that needs to run?).

**Comment by clason** [comment]
Indeed. (Although right now, the required workflows do not distinguish between the kind of PR, so it _does_ need to be run at least once in order to be able to merge it.)

**Comment by theHamsta** [comment]
@clason @dundargoc Actions can trigger other actions, however actions performed by a git user not associated to a Github account get ignored for triggered for actions. This is why you manually have to open/close the "update XXX" PR. When we would create a Github user (like Neovim's @marvim  with personal access token in the repo secrets that could trigger actions.

**Comment by dundargoc** [comment]
The CI logs show that swift highlights take around 370 seconds, which is around >90% of the entire job (ubuntu-latest, gcc, stable). This job is also the one that usually takes the longest time. Anyone know why swift takes such a long time compared to all of the other languages?

**Comment by clason** [comment]
Yes. For reasons. (Technical reasons that were fixed in tree-sitter; unfortunately, Neovim stable still has that old version without that fix. Nightly is _much_ faster, so this will cease to be an issue when 0.7 is released.)

**Comment by dundargoc** [comment]
Thanks for the explanation :+1: 

This is good news, means that this particular bottleneck will solve itself in due time then.

**Comment by clason** [comment]
Of course, we could skip that parser on stable if we want to, at the cost of a slight increase in complexity of the test script/workflow.

**Comment by clason** [comment]
That has pretty much been addressed (thanks @dundargoc !)
